sender/firehose_sender.go

```diff
@@ -0,0 +1,138 @@
+	// DeployEnv is the name of the runtime environment ("development" or "production")
+	// It is used in the decoder to inject an environment into logs.
+	DeployEnv string
+	// FirehoseRegion the region in which the firehose exists
+	FirehoseRegion string
+	// StreamName is the firehose stream name
+	StreamName string
+	// StringifyNested will take any nested JSON objects and send them as strings instead of JSON objects.
+	StringifyNested bool
+	// RenameESReservedFields will rename any field reserved by ES, e.g. _source, to kv__<field>, e.g. kv__source.
+	RenameESReservedFields bool
+	// MinimumTimestamp will reject any logs with a timestamp < MinimumTimestamp
+	MinimumTimestamp time.Time
+}
+
+// NewFirehoseSender creates a FirehoseSender
+func NewFirehoseSender(config FirehoseSenderConfig) *FirehoseSender {
+	f := &FirehoseSender{
+		streamName:             config.StreamName,
+		deployEnv:              config.DeployEnv,
+		stringifyNested:        config.StringifyNested,
+		renameESReservedFields: config.RenameESReservedFields,
+		minimumTimestamp:       config.MinimumTimestamp,
+	}
+
+	awsConfig := aws.NewConfig().WithRegion(config.FirehoseRegion).WithMaxRetries(10)
+	sess := session.Must(session.NewSession(awsConfig))
+	f.client = firehose.New(sess)
+
+	return f
+}
+
+// ProcessMessage processes messages
+func (f *FirehoseSender) ProcessMessage(rawlog []byte) ([]byte, []string, error) {
+	fields, err := decode.ParseAndEnhance(
+		string(rawlog), f.deployEnv, f.stringifyNested,
+		f.renameESReservedFields, f.minimumTimestamp,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	msg, err := json.Marshal(fields)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// add newline after each record, so that json objects in firehose will apppear one per line
+	msg = append(msg, '\n')
+
+	return msg, []string{f.streamName}, nil
+}
+
+func (f *FirehoseSender) sendRecords(batch [][]byte, tag string) (
+	*firehose.PutRecordBatchOutput, error,
+) {
+	awsRecords := make([]*firehose.Record, len(batch))
+	for idx, log := range batch {
+		awsRecords[idx] = &firehose.Record{Data: log}
+	}
+
+	return f.client.PutRecordBatch(&firehose.PutRecordBatchInput{
+		DeliveryStreamName: &tag,
+		Records:            awsRecords,
+	})
+}
+
+// SendBatch sends batches to a firehose
+func (f *FirehoseSender) SendBatch(batch [][]byte, tag string) error {
+	res, err := f.sendRecords(batch, tag)
+	if err != nil {
+		return kbc.CatastrophicSendBatchError{err.Error()}
+	}
+
+	retries := 0
+	delay := 250
+	for *res.FailedPutCount != 0 {
+		log.WarnD("retry-filed-records", logger.M{
+			"stream": tag, "failed-record-count": *res.FailedPutCount, "retries": retries,
+		})
+
+		time.Sleep(time.Duration(delay) * time.Millisecond)
+
+		retryLogs := [][]byte{}
+		for idx, entry := range res.RequestResponses {
+			if entry != nil && entry.ErrorMessage != nil && *entry.ErrorMessage != "" {
+				log.ErrorD("failed-record", logger.M{"stream": tag, "msg": &entry.ErrorMessage})
```

Comment:
I think we may want to remove the error log line per failed record. In the worst case, this would log every log, leading to a cascading problem.

Otherwise, we should rate-limit it somehow, or add some sanity checks (don't log if 100% failures)